ee/clickhouse/queries/session_recordings/clickhouse_session_recording_list.py

```diff
@@ -1,119 +1,234 @@
-from typing import Any, Dict, List, NamedTuple, Tuple
+from datetime import timedelta
+from typing import Any, Dict, List, NamedTuple, Set, Tuple, Union
 
 from ee.clickhouse.client import sync_execute
 from ee.clickhouse.models.action import format_entity_filter
-from ee.clickhouse.models.entity import get_entity_filtering_params
 from ee.clickhouse.models.property import parse_prop_clauses
+from ee.clickhouse.models.util import PersonPropertiesMode
 from ee.clickhouse.queries.person_query import ClickhousePersonQuery
+from posthog.constants import TREND_FILTER_TYPE_ACTIONS
 from posthog.models.entity import Entity
-from posthog.models.filters.session_recordings_filter import SessionRecordingsFilter
-from posthog.queries.session_recordings.session_recording_list import (
-    EventsQueryWithAggregateClausesSQL,
-    SessionRecordingList,
-    SessionRecordingQueryResult,
-)
+from posthog.queries.session_recordings.session_recording_list import SessionRecordingList, SessionRecordingQueryResult
 
 
 class EventFiltersSQL(NamedTuple):
-    event_select_clause: str
-    event_where_clause: str
     aggregate_select_clause: str
-    aggregate_where_clause: str
+    aggregate_having_clause: str
+    where_conditions: str
     params: Dict[str, Any]
 
 
 class ClickhouseSessionRecordingList(SessionRecordingList):
-    _recording_duration_select_statement = (
-        "dateDiff('second', toDateTime(MIN(timestamp)), toDateTime(MAX(timestamp))) as duration,"
-    )
-    _recording_full_snapshot_select_statement = "COUNT((JSONExtractInt(snapshot_data, 'type') = 2 OR JSONExtractBool(snapshot_data, 'has_full_snapshot')) ? 1 : NULL) as full_snapshots"
-    _session_recording_event_table = "session_recording_events"
-    _session_recording_select_statements = """
-            any(session_recordings.start_time) as start_time,
-            any(session_recordings.end_time) as end_time,
-            any(session_recordings.duration) as duration,
-            any(filtered_events.distinct_id) as distinct_id
-    """
-
-    _event_query = """
+    _session_recordings_query_with_entity_filter: str = """
+    SELECT
+        session_recordings.session_id,
+        any(session_recordings.start_time) as start_time,
+        any(session_recordings.end_time) as end_time,
+        any(session_recordings.duration) as duration,
+        any(session_recordings.distinct_id) as distinct_id,
+        arrayElement(groupArray(current_url), 1) as start_url,
+        arrayElement(groupArray(current_url), -1) as end_url
+        {event_filter_aggregate_select_clause}
+    FROM (
         SELECT
+            distinct_id,
+            event,
+            team_id,
             timestamp,
-            distinct_id
-            {event_filter_event_select_clause}
+            properties,
+            JSONExtractString(properties, '$current_url') as current_url
```

Comment:
Both this line and above - we should also make use of materialized columns here. This can speed up queries by a factor of 2-3 if done correctly.

See ee/clickhouse/queries/trends/trend_event_query.py for an example - it uses `column_optimizer` to figure out what columns we need to query and queries them.

Reply:
I think I got this working, but I would love another set of eyes on my work. I'm a bit confused about what `allow_denormalized_props` means.